File: day_02/python3/day_2_1.py
from sys import stdin
from operator import add, mul

prog = list(map(int, stdin.read().split(",")))
op = {1: add, 2: mul}

prog[1] = 12
prog[2] = 2

for pos in range(0, len(prog), 4):
    a, b, c, d = prog[pos:pos+4]
    if a == 99: break
    prog[d] = op[a](prog[b], prog[c])

print(prog[0])

File: day_02/haskell/day_2_1.hs
import Data.List.Split
import Data.Vector as V

main :: IO ()
main = interact (show . run . V.map read . fromList . splitOn ",")

run :: Vector Int -> Int
run prog = V.head $ run' (prog // [(1, 12), (2, 2)]) 0
run' :: Vector Int -> Int -> Vector Int
run' prog i
    | opc == 99 = prog
    | otherwise = run' (prog // [(out, op prog opc a b)]) (i + 4)
  where
    opc = prog ! i
    a   = prog ! (i + 1)
    b   = prog ! (i + 2)
    out = prog ! (i + 3)

op :: Vector Int -> Int -> Int -> Int -> Int
op prog 1 a b = (prog ! a) + (prog ! b)
op prog 2 a b = (prog ! a) * (prog ! b)
